optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/calendar/TwoDayView.java

```diff
@@ -0,0 +1,1013 @@
+    private Pagination pagination;
+    private SimplePager pager;
+    private ListDataProvider<Collection<D>> dataProvider = new ListDataProvider<>();
+    private SelectionModel<? super Collection<D>> selectionModel;
+    private int rangeStart, rangeEnd;
+    private int totalDisplayedSpotSlots;
+    LocalDateTime baseDate;
+    LocalDateTime currDay;
+    LocalDateTime hardStartDateBound;
+
+    @Override
+    public LocalDateTime getHardStartDateBound() {
+        return hardStartDateBound;
+    }
+
+    @Override
+    public void setHardStartDateBound(LocalDateTime hardStartDateBound) {
+        this.hardStartDateBound = hardStartDateBound;
+        if (getViewStartDate().isBefore(hardStartDateBound)) {
+            setDate(hardStartDateBound);
+        }
+        updateScrollBar();
+    }
+
+    @Override
+    public LocalDateTime getHardEndDateBound() {
+        return hardEndDateBound;
+    }
+
+    private void updateScrollBar() {
+        double oldPos = (scrollBar.getHorizontalScrollPosition() + 0.0) / scrollBar
+                .getMaximumHorizontalScrollPosition();
+        scrollBar.setWidth(Math.round(screenWidth) + "px");
+
+        if (null != hardStartDateBound && null != hardEndDateBound) {
+            int daysBetween = (int) ((hardEndDateBound.toEpochSecond(ZoneOffset.UTC) - hardStartDateBound.toEpochSecond(
+                    ZoneOffset.UTC)) / (60 * 60 * 24));
+            scrollBar.setScrollWidth((int) Math.round(screenWidth * ((daysBetween + 0.0) / daysShown)));
+            scrollBar.setHorizontalScrollPosition((int) Math.round(oldPos * scrollBar
+                    .getMaximumHorizontalScrollPosition()));
+        } else {
+            scrollBar.setScrollWidth(0);
+        }
+    }
+
+    @Override
+    public void setHardEndDateBound(LocalDateTime hardEndDateBound) {
+        this.hardEndDateBound = hardEndDateBound;
+        if (getViewEndDate().isAfter(hardEndDateBound)) {
+            setDate(hardEndDateBound.minusDays(daysShown));
+        }
+        updateScrollBar();
+    }
+
+    LocalDateTime hardEndDateBound;
+
+    double mouseX, mouseY;
+    double localMouseX, localMouseY;
+    double screenWidth, screenHeight;
+    double dragStartX, dragStartY;
+    double widthPerMinute, spotHeight;
+    int totalSpotSlots;
+    int daysShown;
+    int editMinuteGradality;
+    int displayMinuteGradality;
+
+    public int getDaysShown() {
+        return daysShown;
+    }
+
+    public void setDaysShown(int daysShown) {
+        this.daysShown = daysShown;
+        updateScrollBar();
+    }
+
+    public int getEditMinuteGradality() {
+        return editMinuteGradality;
+    }
+
+    public void setEditMinuteGradality(int editMinuteGradality) {
+        this.editMinuteGradality = editMinuteGradality;
+    }
+
+    public int getDisplayMinuteGradality() {
+        return displayMinuteGradality;
+    }
+
+    public void setDisplayMinuteGradality(int displayMinuteGradality) {
```

Comment:
displayMinuteGradality is currently specific to this view; it controls the bars drawn on the screen (so 60 -> 1 bar = 1 hour, 120 -> 1 bar = 2 hour). I think it is closer to view than presenter, since displayMinuteGradality is only used for drawing. editMinuteGradality probably can be moved to presenter since it does directly relate to the domain. I can move these kind of variables (view parameters) to an internal state class kept in the instance of the view.

Reply:
> displayMinuteGradality is currently specific to this view

If that's the case, then this view should have it's own presenter that holds this state.

> I think it is closer to view than presenter, since displayMinuteGradality is only used for drawing.

What makes code related to view or presenter doesn't have to do with how UI-related it is. A view is meant to be a thin wrapper around the UI primitives of your runtime environment. In principle, every method in your view should map directly to part of the API of your UI primitives; ideally as a single method call, and definitely without any intermediate calculations or control flow.

Display granularity is not a UI primitive of the web-browser, it is a concept specific to your component. In theory, if you do MVP right, you could port your entire application just by rewriting views (assuming the origin and destination environments had similar semantics).